.appveyor.yml

```diff
@@ -13,10 +13,17 @@ branches:
 
 environment:
   matrix:
-    - COMPILER: MINGW
-      TASK: r-package
-    - COMPILER: MSVC
-      TASK: r-package
+    - TASK: r-package
+      R_VERSION: 3.6
+      COMPILER: MINGW
+      TOOLCHAIN: MINGW
+    - TASK: r-package
+      R_VERSION: 4.0
+      COMPILER: MINGW
+      TOOLCHAIN: MSYS
+    - TASK: r-package
+      R_VERSION: 4.0
+      COMPILER: MSVC
```

Comment:
It's too much I think! Let's keep all possible combinations for #2353. For now it should be enough 3 R jobs: 2 here and 1 at Azure.

Reply:
Sorry, what do you mean by "keep all possible combinations for #2353"?

Do you mean "if you want more concurrent jobs, write them in GitHub Actions"? I'm happy to try that in a followup PR, but it will take some time.

These are the minimum relevant things to cover for the R package on Windows

* R version
    - 3.6.x
    - 4.0.x
* toolchain:
    - MINGW
    - MSYS
    - Visual Studio
        - 14 (2015)
        - 15 (2017)
        - 16 (2019)

That is 10 different combinations, conservatively. I know we can't test all of them, but I feel we should have at least 4 Windows R builds:

* Visual Studio, 4.0.x
* Visual Studio, 3.6.x
* non-Visual Studio (MinGW), 3.6.x
* non-Visual Studio (MSYS), 4.0.x

If you still disagree and think we should cap it at 3 for now, I'm I guess I'd remove the Visual Studio + R 3.6 one

* Visual Studio, 4.0.x
* ~Visual Studio, 3.6.x~
* non-Visual Studio (MinGW), 3.6.x
* non-Visual Studio (MSYS), 4.0.x

Reply:
To add more context, I understand what you're saying with https://github.com/microsoft/LightGBM/pull/3065#discussion_r428251729 that maybe there is too much duplication in Mac / Linux builds across Travis + Azure, but for Windows I think it's important to test all four of the cases I mentioned above, since the toolchain changed between major versions

Reply:
Yeah, I mean that more covering test cases can be added for GitHub Actions to not increase the overall CI time dramatically. 

> - Visual Studio, 4.0.x
> - Visual Studio, 3.6.x

Can you please clarify what will be the difference in these test cases?

Reply:
Got it! Ok yes I agree with that.

> Can you please clarify what will be the difference in these test cases?

* `Visual Studio, 4.0.x` = Visual Studio, R 4.0, and `Rtools40` (would have caught #3064 )
* `Visual Studio, 3.6.x` = Visual Studio, R 3.6.3, and `Rtools35` (would catch if this PR breaks backwards compatibility with older R versions)

Reply:
But after this PR is merged, all logic for Visual Studio will be the same (trial projects generation, `.def` file generation). Difference will be only in PATH content, but all issues with R-version-specific paths can be caught by MinGW/MSYS builds. Are my statements correct or I'm missing something?

Reply:
Even though LightGBM code would look the same, these types of things would not be caught if we don't have a build with Visual Studio + R 3.6.x:

* a linker flag is added to `CMakeLists.txt` in an `if(MSVC)` block that works when linking to the `R.lib` for R 4.0 but not to the `R.lib` for R 3.6.3
* C++ code is added in a block guarded by `#ifdef _MSC_VER` which uses a function that R prohibits (like `abort`), or which conflicts with some other symbol being read in from `R.h` or the other R headers

Basically the existence of `if(MSVC)` in `CMakeLists.txt` and `#ifdef _MSCV_VER` in our C++ code opens the possibility of something being incompatible between a `LightGBM` compiled with MSVC and R 3.6.3.

We know as of this moment that there are no such issues because we have a CI build that tests that combination. If we're very worried about CI time, I am ok dropping the build that is "Visual Studio + R 3.6.3" because of that, but it does mean we're losing some coverage of the issues I mentioned above.

Reply:
Ah, the second point about conflicts with different (due to major version difference) R headers makes a lot of sense! TBH, I didn't get the first one.

OK, keeping in mind that all R jobs will be held by GitHub Actions eventually (I hope), maybe at least we should balance these 4 jobs better for now? Appveyor is extremely slow, you know.

Reply:
since I made so much progress on #3116 , I'm confident we'll have GitHub Actions in a few days. We could get that reviewed / merged first, then I could change this PR to only add GitHub Actions jobs!

20  concurrent tasks mean we could comfortably add all the different R combinations

Reply:
Amazing!

Reply:
I've moved all Linux and Mac jobs to GitHub Actions in #3116 , and took your advice to balance Windows jobs better for now (since adding Windows jobs to GitHub Actions was blocked by https://github.com/PowerShell/PowerShell/issues/12823, from https://github.community/t/powershell-steps-fail-nondeterministically/115496/9)

As of this PR, we'll have two R builds on AppVeyor and two on Azure

* AppVeyor, R 4.0, MSYS2
* AppVeyor, R 4.0, Visual Studio 2019
* Azure, R 3.6, MINGW
* Azure, R 3.6, Visual Studio 2017